### PR TITLE
Adds more UI/UX verbosity to odo push.

### PR DESCRIPTION
**Pull request by cdrage**
**What kind of PR is this?**

/kind epic
/kind feature

Phew. So there's a lot of changes here. Let's go through them.

- When pushing `odo push`, more verbosity is shown such as the pod
status as well as the deployment status. This looks like the following
below:

```sh
Pushing to component nodejs-nodejs-ex-nhcg of type local
 ┌ Deploying...
 ├─ ◓ Pod: Pushing to cluster [7s]
 └─ ◓ DeploymentConfig: Requested Replicas (0) | Available Replicas (0) [7s]
```

When completed:

```sh
▶ odo push -f
Validation
 ✓  Checking component [196ms]

Configuration changes
 ✓  Retrieving component data [234ms]
 ✓  Applying configuration [202ms]

Pushing to component nodejs-nodejs-ex-nhcg of type local
 ┌ Deploying...
 ├─ ✓ Pod: Running [12s]
 └─ ✓ DeploymentConfig: Requested Replicas (1) | Available Replicas (1) [12s]
 ✓  Syncing files to the component [713ms]
 ✓  Building component [15s]
 ✓  Changes successfully pushed to component
```

- A lot more concurrency (see all the new go routines, etc.)
- Tests have been added to cover all major scenarios when pushing a pod

**What does does this PR do / why we need it**:

Why? Because it adds *a lot* more verbosity and output as to what odo is
doing to the background.

**Which issue(s) this PR fixes**:

Doesn't `close` https://github.com/openshift/odo/issues/2244 but
"enhances" it and adds at least more verbosity as to what has failed.

**How to test changes / Special notes to the reviewer**:

```sh
git clone https://github.com/openshift/nodejs-ex
odo push
```

### CODE REVIEW

**Comment by cdrage** [comment]
# TODOs

- [ ] Possible integration tests (not really necessary since all the `odo push` already covers it
- [ ] Update vendoring (add cdrage/uilive)
- [ ] Add a bunch more comments
- [ ] Add updated asciinema gif / to show what's happening
- [x] Windows support

File: glide.yaml [hunk 1]
```diff
@@ -64,3 +64,5 @@ import:
   version: fa7b53cdfc9105c70f134574002f406232921437
 - package: github.com/spf13/afero
   version: v1.2.2
+- package: github.com/cdrage/uilive
```
Comment (kadel) [inline glide.yaml]: why forked version?

Comment (cdrage) [reply glide.yaml]: Needed to implement a few open PR's that the maintainer has yet to merge in: https://github.com/gosuri/uilive/pull/29 so using a forked version with those changes.

**Comment by cdrage** [comment]
Ping @girishramnani @dharmit @kadel @dharmit @mohammedzee1000 and all! Ready for review.

**Comment by girishramnani** [comment]
how does this look on a windows machine?

**Comment by cdrage** [comment]
@girishramnani Test it and find out :) Works well in Powershell + CMD!

**Comment by kadel** [comment]
I don't know the output looks a little bit weird to me :-(

```
Pushing to component frontend of type local
 ✓  Checking files for pushing [182ms]
 ┌ Deploying...
 ├─ ✓ Pod: Running [8s]
 └─ ✓ DeploymentConfig: Requested Replicas (1) | Available Replicas (1) [8s]
 ✓  Syncing files to the component [10s]
 ✓  Building component [32s]
 ✓  Changes successfully pushed to component

```
`Deploying ...` doesn't fit there :-(
Everything else has checkmarks and this start with line. Also it looks like it belongs under ` ✓  Checking files for pushing [182ms]` which doesn't make sense :-(

Could we replace `Deploying ...` with `Preparing component ...` ? And have a spinner and checkmark at the beginning of the line?

File: pkg/occlient/occlient.go [hunk 2]
```diff
@@ -1708,6 +1708,7 @@ func (c *Client) WaitForBuildToFinish(buildName string) error {
 //	waitCond: Function indicating when to consider dc rolled out
 // Returns:
 //	Updated DC and errors if any
+// TODO: Move this function into a separate file! (wait.go?), also, why isn't there a go routine here?
```
Comment (kadel) [inline pkg/occlient/occlient.go]: Why would you need a go routine?

File: pkg/occlient/monitor.go [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+)
+
+// Error channels for both the pod and deploymentconfig
+var deploymentConfigErrorChannel chan error
+var podErrorChannel chan error
+
+// Main wait group for both go routines.
+var wg sync.WaitGroup
+
+// waitForDD is a go routine to wait for the DeploymentConfig to successfully complete / come up
```
Comment (kadel) [inline pkg/occlient/monitor.go]: typo `s/waitForDD/waitForDC`.
Is it really a go routine?

File: pkg/occlient/monitor.go [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+func (c *Client) waitForDC(name string, deploymentConfigSpinner *fidget.Spinny) {
```
Comment (kadel) [inline pkg/occlient/monitor.go]: Why you couldn't use already existing `occlient.WaitAndGetDC` function? It has almost the same logic.

File: pkg/occlient/monitor.go [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+		} else {
+			deploymentConfigErrorChannel <- errors.New("unable to convert event object to DeploymentConfig")
+			break
+		}
+	}
+
+}
+
+// waitForPod waits for the pod to successfully complete / come up
+func (c *Client) waitForPod(podSelector string, podSpinner *fidget.Spinny) {
```
Comment (kadel) [inline pkg/occlient/monitor.go]: again, why new function if there already is `occlient.WaitAndGetPod`?

File: pkg/occlient/monitor.go [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+
+		} else {
+			podErrorChannel <- errors.New("unable to convert event object to Pod")
+			break
+		}
+	}
+	podErrorChannel <- errors.New("unable to convert event object to Pod")
+}
+
+// WaitForEverything will wait for EVERYTHING to be RUNNING! (replacement for "WaitAndGetPod" since this far more verbose
```
Comment (kadel) [inline pkg/occlient/monitor.go]: What is EVERYTHING?

File: pkg/occlient/monitor.go [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+// and elaborate..
+// TODO:
+// Use `kubectl get events -w` for LAST occuring event? Also, we shouldn't really be displaying this if we're redeploying? Else new
+// PVC's need to be created, etc.
+// For --git build should we output the same stuff? Got to output builconfig too?
+// Secrets
+// PVC?
+// buildconfig?
+func (c *Client) WaitForEverything(podSelector string, name string, timeout time.Duration) error {
```
Comment (kadel) [inline pkg/occlient/monitor.go]: what is `name`? - name of what?

Comment (kadel) [reply pkg/occlient/monitor.go]: If name is DeploymentConfig name, than you don't need `podSelector` as you can get it from DeploymentConfig.

**Comment by girishramnani** [comment]
> # TODOs
> * [ ]  Possible integration tests (not really necessary since all the `odo push` already covers it
> * [ ]  Update vendoring (add cdrage/uilive)
> * [ ]  Add a bunch more comments
> * [ ]  Add updated asciinema gif / to show what's happening
> * [x]  Windows support

please update this list

**Comment by cdrage** [comment]
> > # TODOs
> > 
> > * [ ]   Possible integration tests (not really necessary since all the `odo push` already covers it
> > * [ ]   Update vendoring (add cdrage/uilive)
> > * [ ]   Add a bunch more comments
> > * [ ]   Add updated asciinema gif / to show what's happening
> > * [x]   Windows support
> 
> please update this list

Please review :)

**Comment by mik-dass** [comment]
```
✓  Checking files for pushing [2ms]
 ┌ Deploying...
 ├─ ✓ Pod: Running [50s]
 └─ ✓ DeploymentConfig: Requested Replicas (1) | Available Replicas (1) [50s]
 ✓  Syncing files to the component [19s]
 ✓  Building component [17s]
 ✓  Changes successfully pushed to component
```
@kadel @cdrage I think the output might confuse the user as terms like `pod`, `deploymentconfig` etc are used. Also this doesn't seem to work for git components.

**Comment by girishramnani** [comment]
I am putting a hold on this PR as it seems to make the output a bit more confusing.
/hold

**Comment by cdrage** [comment]
Going to close this for now... this was a task that I really over-engineered.

A new implementation will go up that's simpler and doesn't have as many moving parts.

**Pull request closed without merging**